### PR TITLE
EOS-17966: Implemented event manager publish method

### PR DESCRIPTION
**Pull request by indrajitzagade**
## Problem Statement
<pre>
  <code>
  Story Ref (if any):
    EOS-17966: Implemented event manager publish method
  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
  Yes (Tested by running the publisher test directly without building/installing rpm)
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Your Problem decription here...
  </code>
</pre>
## Solution
<pre>
  <code>
    Implement interface to publish HA events on the message bus
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
    Ran the publisher test successfully.
  </code>
</pre>

### CODE REVIEW

**Comment by indrajitzagade** [comment]
To avoid duplicate efforts, waiting for PR #509 and #510 to get merged. Some update to current PR will be needed after this merge.

File: ha/core/event_manager/event_manager.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            if self._confstore.key_exists(component_list_key):
+                component_list_key_val = self._confstore.get(component_list_key)
+                _, value = component_list_key_val.popitem()
+                component_list = json.loads(value)
+            for component in component_list:
+                producer_id = const.EVENT_MGR_PRODUCER_ID.replace("<component_id>", component)
+                message_type_key = const.EVENT_MGR_MESSAGE_TYPE_KEY.replace("<component_id>", component)
+                message_type_key_val = self._confstore.get(message_type_key)
+                _, message_type = message_type_key_val.popitem()
+                message_producer = MessageProducer(producer_id = producer_id,
```
Comment (ajay-paratmandali) [inline ha/core/event_manager/event_manager.py:151]: Can we saperate message bus and this code saperate
message bus needed many place so want to centralize it

https://github.com/Seagate/cortx-ha/pull/514/files
created class under util for same reason

File: ha/util/message_bus.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python3
```
Comment (indrajitzagade) [inline ha/util/message_bus.py:1]: This file has been copied from PR #514

File: ha/core/event_manager/event_manager.py (hunk 1 above)
Comment (indrajitzagade) [reply ha/core/event_manager/event_manager.py]: done, using the class available in PR #514

**Review by ajaysrivas — APPROVED** [review]

**Review by ajay-paratmandali — APPROVED** [review]

**Comment by ajay-paratmandali** [comment]
Please resolve conflict 
@indrajitzagade

**Comment by indrajitzagade** [comment]
> 
> 
> Please resolve conflict
> @indrajitzagade

Resolved @ajaysrivas

File: ha/core/event_manager/event_manager.py [hunk 3]
```diff
@@ -333,19 +334,37 @@ def get_events(self, component : str) -> list:
...
         Publish event.
         Args:
-            component (str): Component name.
             event (RecoveryActionEvent): Action event.
         """
-        pass
+        try:
+            component_list = []
+            # Run through list of components subscribed for this event and send event to each of them
+            component_list_key = f'{const.EVENT_KEY}/{event.resource_type}/{event.event_type}'
```
Comment (ajay-paratmandali) [inline ha/core/event_manager/event_manager.py:346]: can we have _prepare_key method to get this keys

Comment (indrajitzagade) [reply ha/core/event_manager/event_manager.py:346]: If you mean the _prepare_key() method in system health module, then I guess lets not it now.
We will have to make it generic before use.

File: ha/core/event_manager/event_manager.py [hunk 4]
```diff
@@ -333,19 +334,37 @@ def get_events(self, component : str) -> list:
...
+            if self._confstore.key_exists(component_list_key):
```
Comment (ArchanaLimaye) [inline ha/core/event_manager/event_manager.py:347]: can skip this check.
 _confstore.get will return "None" if key is not found and that "None" can be used to skip the publish call

Comment (ArchanaLimaye) [inline ha/core/event_manager/event_manager.py:347]: can skip this check.
_confstore.get will return "None" if key is not found and that "None" can be used to skip the publish call

File: ha/core/event_manager/const.py [hunk 5]
```diff
@@ -27,8 +27,11 @@ class SUBSCRIPTION_LIST(enum.Enum, metaclass=EnumListMeta):
...
     TEST = "test"
 
+ACTION_EVENT_VERSION = "2.0"
 COMPONENT_KEY = 'events/subscribe'
 EVENT_KEY = 'events'
 HA_MESSAGE_BUS_ADMIN = "ha_admin"
+EVENT_MGR_PRODUCER_ID = "ha_event_manager_<component_id>"
 EVENT_MGR_MESSAGE_TYPE = "ha_event_<component_id>"
 EVENT_MGR_MESSAGE_TYPE_KEY = "message_type/<component_id>"
+EVENT_COMPONENT_LIST = "events/<event_name>"
```
[the remark this answers is not in the document]
Comment (indrajitzagade) [reply ha/core/event_manager/const.py]: Good catch, deleted

File: ha/core/event_manager/event_manager.py (hunk 4 above)
Comment (indrajitzagade) [reply ha/core/event_manager/event_manager.py]: Done

**Review by ArchanaLimaye — APPROVED** [review]

**Pull request merged**